redash/models/__init__.py

```diff
@@ -1036,6 +1038,28 @@ def render_template(self, template):
         else:
             result_value = None
 
+        # Generate an HTML table for a generic display of query data
+        html = io.StringIO("")
+        html.write("<table class=\"query-result\">")
+        html.write("<thead>\n<tr>")
+        columns = [col['name'] for col in data["columns"]]
+        for column in columns:
+            html.write(f"<th>{column}</th>")
+        html.write("</tr>\n</thead>")
+
+        html.write("<tbody>")
+        for row in data["rows"]:
+            html.write(f"\n<tr>")
+            for col in columns:
+                if row[col] is None:
+                    html.write("<td></td>")
+                else:
+                    html.write(f"<td>{row[col]}</td>")
+            html.write(f"</tr>")
+        html.write("\n</tbody>")
+        html.write("</table>")
+        html.seek(0)
+
```

Comment:
I think these should be implemented in the template rather than here.

Reply:
I'm not that great at Python yet, so I'm avoiding doing technical reviews of the Python bits for now.

If it was Go based, then I'd be good... but it's not. :wink:

Reply:
Both @wlach and @guidopetri seem decent at Python, so maybe ask them? :smile:

Reply:
I don't think there's an elegant way of doing this in the template, so I'm good with it like this. We can always improve upon it later.

Reply:
I do not think it is possible to move this logic to the template because the rows are provided as a dictionary.
This is as close as I could get:

```html
<table>
  <thead>
    <tr>
    {{#QUERY_RESULT_COLS}}
    <th>{{friendly_name}}</th>
    {{/QUERY_RESULT_COLS}}
  </tr>
  </thead>

{{#QUERY_RESULT_ROWS}}
  <tr>
    <td>{{Volume}}</td>
    <td>{{Max Depth}}</td>
    <td>{{Average Depth}}</td>
  </tr>
{{/QUERY_RESULT_ROWS}}
</table>
```

(Notice that `Volume`, `Max Depth`, `Average Depth` are column names and are specific to this query.)

If each row was a list we could write an inner loop for each column, so perhaps the right approach is to provide another view on the data that is more in line with the way Mustash works.

Reply:
I don't think passing an HTML string to a template is a good approach. The purpose of a template is customization, and providing the rendered result to the template is an anti-pattern.